docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+  - MCUBoot does the updates and authentication of all image, and if all checks
+    pass, it boots the image from primary slot of the 0th position.
+  - Other image in the boot chain is started by previous image.
+  - This proposal extends the current TLV manifest format. IETF manifest format
+    was out of scope.
+  - Revert feature of MCUBoot is also supported for multiple images.
+
+# Terminology and slot layout
+  - Header: image header
+  - Payload: image itself
+  - Manifest: contains TLV entries(image hash, key hash, signature, etc.)
+  - Image trailer:
+    - Swap status: progress info during swapping
+    - Swap state:  `struct image_trailer`
+
+The order of the items described above also represents the memory layout of a
+single slot. There can be some gaps (which is padded) between the following
+elements:
+  - header - payload: vector table must be on aligned address
+  - manifest - image trailer: image is smaller then memory slot
+
+```
+  +----------------------+
+  | Header               |   <- struct image_header
+  +----------------------+
+  | Payload              |   <- image
+  +----------------------+
+  | Manifest             |   <- TLV entries
+  +----------------------+
+  | Unused space in slot |   <- Filled with 0xFF
+  +----------------------+
+  | Trailer              |
+  | +------------------+ |
+  | | Swap status      | |   <- Progress info
+  | +------------------+ |
+  | | Swap state       | |   <- struct image_trailer
+  | +------------------+ |
+  +----------------------+
+```
+# Describing dependency
+As David Brown suggests the dependency between images can be described with
+additional TLV entries in the manifest.
+``` c
+#define IMAGE_TLV_DEPENDENCY  0x40   /* image depends on other image */
+
+#define IMAGE_CNT 2 /* Platform dependent, how many individual images are there
+                     */
+struct image_version {
+   uint8_t iv_major;
+   uint8_t iv_minor;
+   uint16_t iv_revision;
+   uint32_t iv_build_num;
+};
+
+struct image_dependency {
+    uint32_t image_id;  /* For simplicity could be just numbering from 0 to
+                         * IMAGE_CNT
+                         */
+    struct image_version image_min_version;  /* Indicates at minimum which
+                                              * version of which firmware must
+                                              * be available to satisfy
+                                              * compliance
+                                              */
+};
+```
+
+*Note: IETF manifest proposal suggest to use exact hash value instead of version
+number to describe dependency*
+
+In the manifest section can be more then one dependency entry, but in practice
+if the platform only supports two individual images then there can be maximum
+one entry which reflects to the other image. I.e.: Non-secure image can be
+dependent on secure image and vice versa.
+
+Dependency is a mutual relation. If one image depends on an other then that
```

Comment:
When is this true? If an S firmware is dependent on an NS firmware, I would consider that a security issue. Also, e.g. a network stack doesn't care who is using it.

Reply:
Why a security issue, pls. elaborate a bit? 

On armv8M architecture some functionalities are separated between S ans NS world. But S world provide some service to the NS. They are communicating with a defined API, which by default introduce a dependency between them:
- If API provided by S has changed then NS also must be updated. NS which supports older version of API might cannot collaborate with S, which supports newer API version.
- Also the SG(secure gateway) instruction are placed to a fixed place as part of the S image. If they are moved in memory between S version then a new NS image must be compiled. 

But you are right it can be the case when S and NS image can be updated independently. If there is no change in the interface, only internal changes were made.

Application also depends on network stack if its API changed.

Reply:
You are right, S firmwares can safely depend on NS firmwares, though of course care must be taken whenever data is provided by the NS.

To be clear, I disagree with the following paragraph as a blanket statement

> Dependency is a mutual relation. If one image depends on an other then that
other image also depend on the first one. Therefore in case of dependency there
must be one dependency entry in both images manifest in order to ensure proper
updates.

To me it looks like you're saying that, as a rule, all dependencies shall be reciprocated. This doesn't make sense to me.

I'm not saying circular dependencies don't have valid use cases. I'm saying there's a difference between one-way and two-way dependencies, and this distinction should be leveraged when it is useful to do so.

A natural (to me) interpretation of versions is that the FW version strictly means "API version", meaning that a FW has a dependency on another FW iff it actively calls on that FW. Circular dependencies have only a minor role here IMO, in cases where two FWs actively use each other's API. A network stack is a clear case of one-way dependency, but I realize that in ARMv8 scenarios there is a larger chance of circular dependencies happening.